Xformer SDP forward/backward kernel

drisspg opened the pull request:
# Summary
Include xformer kernel code and make header updates to successfully build. Need to update the kernel calling code and dispatch system to clean this up.

drisspg commented on aten/src/ATen/native/transformers/cuda/attention.cu:
```diff
@@ -480,17 +544,40 @@ std::tuple<Tensor, Tensor> flash_attention_helper_dense_unpacked(
+    const Tensor& value){
+  // Query -> Query(Batch x Q_seq_len x Num_heads x Dim_per_head)
+  // Key   -> Key(Batch x KV_seq_len x Num_heads x Dim_per_head)
+  // Value -> Value(Batch x KV_seq_len x  Num_heads x Dim_per_head)
```
very minimal but still inlining this to the case block would be a little too big

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -179,6 +204,9 @@ inline SDPBackend select_sdp_backend(sdp_params kernel_params) {
   if (use_flash_attention(kernel_params, !print_debug)) {
     return SDPBackend::flash_attention;
   }
+  if (use_mem_efficient_attention(kernel_params, !print_debug)){
```
Problem with this design is how print debug statements. This works because the constraints are strictly less than for mem_effic than flash but if we add another kernel this doesn't make sense.

cpuhrsch replied:
We could have `forceUseSDPFlash` etc. flags that print debug statements if the branch wasn't taken.

drisspg commented:
Context size:
Before:
```
nvidia-smi --query-gpu=utilization.memory,memory.total,memory.free,memory.used --format=csv -i 0
utilization.memory [%], memory.total [MiB], memory.free [MiB], memory.used [MiB]
0 %, 40960 MiB, 39522 MiB, 831 MiB
```

After:
```
nvidia-smi --query-gpu=utilization.memory,memory.total,memory.free,memory.used --format=csv -i 0
utilization.memory [%], memory.total [MiB], memory.free [MiB], memory.used [MiB]
0 %, 40960 MiB, 39518 MiB, 835 MiB
```

cpuhrsch commented:
Ok, so we're adding 4MB or 0.5%. That seems fine, since we're adding fused kernel support for a lot more devices and autograd. cc @ngimel for visibility.

drisspg commented:
@pytorchbot merge

drisspg committed:
Xformer SDP forward/backward kernel (#86157)

Summary:
Include xformer kernel code and make header updates to successfully build. Need to update the kernel calling code and dispatch system to clean this up.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/86157

Differential Revision: D40125774

Pulled By: drisspg

fbshipit-source-id: 8581942556db46a928a6c6229b290693748fc102
